widgets/common/src/filters/FiltersDefinitionForm.tsx

```diff
@@ -28,10 +55,25 @@ const FiltersDefinitionForm: FunctionComponent = () => {
         setRows(latestRows => latestRows.filter(row => row.id !== id));
     }
 
+    function updateRule(id: string, newRule: FilterRule) {
+        setRows(latestRows => latestRows.map(row => (row.id === id ? { id: row.id, rule: newRule } : row)));
+    }
+
+    useUpdateEffect(() => {
+        onChange(rows.map(getFilterRule));
+    }, [rows]);
+
     return (
         <Form>
             {rows.map(row => (
-                <RuleRow key={row.id} removable={rows.length > 1} onRemove={() => removeRule(row.id)} />
+                <RuleRow
+                    key={row.id}
+                    rule={getFilterRule(row)}
+                    removable={rows.length > 1}
+                    onChange={(value: FilterRule) => updateRule(row.id, value)}
```

Comment:
Optional:

Since `onChange` is statically typed, TS can infer the type of `value`, so we can drop it, unless it's there for readability purposes :slightly_smiling_face: 

```suggestion
                    onChange={value => updateRule(row.id, value)}
```

Also, we could name the `value` a bit more concretely :slightly_smiling_face: 

```suggestion
                    onChange={rule => updateRule(row.id, rule)}
```

Reply:
Make sense, I will change it according to your second suggestion.